lib/SILGen/SILGenFunction.cpp

```diff
@@ -795,6 +800,132 @@ void SILGenFunction::emitArtificialTopLevel(Decl *mainDecl) {
+            ctx.getBoolType()->getCanonicalType());
+        falseValue = B.createStruct(moduleLoc, boolType, falseValue);
+      }
+
+      // Create call to taskCreateFlags
+      FuncDecl *taskCreateFlagsFuncDecl = SGM.getTaskCreateFlags();
+      SILFunction *taskCreateFlagsSILFunc = SGM.getFunction(
+          SILDeclRef(taskCreateFlagsFuncDecl, SILDeclRef::Kind::Func),
+          NotForDefinition);
+      SILValue taskCreateFlagsRef =
+          B.createFunctionRefFor(moduleLoc, taskCreateFlagsSILFunc);
+
+      SILType taskPriorityType;
+      {
+        SmallVector<ValueDecl *, 1> decls;
+        concurrencyModule->lookupQualified(
+            concurrencyModule, DeclNameRef(ctx.getIdentifier("TaskPriority")),
+            NL_OnlyTypes, decls);
+        auto decl = cast<NominalTypeDecl>(decls.empty() ? nullptr : decls[0]);
+        taskPriorityType = SGM.Types.getLoweredType(
+            decl->getDeclaredInterfaceType(),
+            taskCreateFlagsSILFunc->getTypeExpansionContext());
+      }
+      SILType optionalTaskPriorityType =
+          SILType::getOptionalType(taskPriorityType);
+
+      SILValue nilTaskPriority =
+          B.createManagedOptionalNone(moduleLoc, optionalTaskPriorityType)
+              .forward(*this);
+
+      SILValue taskFlags = B.createApply(
+          moduleLoc, taskCreateFlagsRef, {},
+          {/*priority*/ nilTaskPriority,
+           /*isChildTask*/ falseValue,
+           /*copyTaskLocals*/ falseValue,
+           /*inheritContext*/ falseValue, /*enqueueJob*/ falseValue,
+           /*addPendingGroupTaskUnconditionally*/ falseValue});
+
+      B.createDeallocStack(moduleLoc, nilTaskPriority);
+
+      // Get main executor
+      FuncDecl *getMainExecutorFuncDecl = SGM.getGetMainExecutor();
+      SILFunction *getMainExeutorSILFunc = SGM.getFunction(
+          SILDeclRef(getMainExecutorFuncDecl, SILDeclRef::Kind::Func),
+          NotForDefinition);
+      SILValue getMainExeutorFunc =
+          B.createFunctionRefFor(moduleLoc, getMainExeutorSILFunc);
+      SILValue mainExecutor =
+          B.createApply(moduleLoc, getMainExeutorFunc, {}, {});
+
+      // Call CreateAsyncTask
+      FuncDecl *builtinDecl = cast<FuncDecl>(getBuiltinValueDecl(
+          getASTContext(), ctx.getIdentifier(getBuiltinName(
+                               BuiltinValueKind::CreateAsyncTask))));
+
+      auto subs = SubstitutionMap::get(builtinDecl->getGenericSignature(),
+                                       {TupleType::getEmpty(ctx)},
+                                       ArrayRef<ProtocolConformanceRef>{});
+
+      SILValue applyMainFunctionRef =
+          B.createPartialApply(moduleLoc, mainFunctionRef, {}, {metatype},
+                               ParameterConvention::Direct_Guaranteed);
+      // Emit the CreateAsyncTask builtin
+      // We can't use emitBuiltinCreateAsyncTask because it doesn't pass the
+      // main metatype into the thunk correctly. :(
```

Comment:
Ignore this comment, it's wrong. Fixed it with the partial apply directly above.